src/core/MOM_open_boundary.F90

```diff
@@ -4648,6 +4744,119 @@ subroutine register_temp_salt_segments(GV, US, OBC, tr_Reg, param_file)
+  type(ocean_OBC_type),       pointer       :: OBC        !< Open boundary structure
+  type(tracer_registry_type), pointer       :: tr_Reg     !< Tracer registry
+  type(param_file_type),      intent(in)    :: param_file !< file to parse for  model parameter values
+  character(len=*),           intent(in)    :: tr_name!< Tracer name
+! Local variables
+  integer :: isd, ied, IsdB, IedB, jsd, jed, JsdB, JedB, nz, nf
+  integer :: i, j, k, n
+  type(OBC_segment_type), pointer :: segment => NULL() ! pointer to segment type list
+  type(tracer_type), pointer      :: tr_ptr => NULL()
+
+  if (.not. associated(OBC)) return
+
+  do n=1, OBC%number_of_segments
+    segment=>OBC%segment(n)
+    if (.not. segment%on_pe) cycle
+    call tracer_name_lookup(tr_Reg, tr_ptr, tr_name)
+    call register_segment_tracer(tr_ptr, param_file, GV, segment, OBC_array=.True.)
+  enddo
+
+end subroutine register_obgc_segments
+
+subroutine fill_obgc_segments(G, GV, OBC, tr_ptr, tr_name)
+  type(ocean_grid_type),      intent(inout) :: G      !< Ocean grid structure
+  type(verticalGrid_type),    intent(in)    :: GV     !< ocean vertical grid structure
+  type(ocean_OBC_type),       pointer       :: OBC    !< Open boundary structure
+  real, dimension(:,:,:),     pointer       :: tr_ptr !< Pointer to tracer field
+  character(len=*),           intent(in)    :: tr_name!< Tracer name
+! Local variables
+  integer :: isd, ied, IsdB, IedB, jsd, jed, JsdB, JedB, n, nz, nt
+  integer :: i, j, k
+  type(OBC_segment_type), pointer :: segment => NULL() ! pointer to segment type list
+
+  if (.not. associated(OBC)) return
+  call pass_var(tr_ptr, G%Domain)
+  nz = G%ke
+  do n=1, OBC%number_of_segments
+    segment => OBC%segment(n)
+    if (.not. segment%on_pe) cycle
+    nt=get_tracer_index(segment,tr_name)
+    if(nt .lt. 0) then
+      call MOM_error(FATAL,"fill_obgc_segments: Did not find tracer "// tr_name)
+    endif
+    isd = segment%HI%isd ; ied = segment%HI%ied
+    jsd = segment%HI%jsd ; jed = segment%HI%jed
+    IsdB = segment%HI%IsdB ; IedB = segment%HI%IedB
+    JsdB = segment%HI%JsdB ; JedB = segment%HI%JedB
+    ! Fill with Tracer values
+    if (segment%is_E_or_W) then
+      I=segment%HI%IsdB
+      do k=1,nz ; do j=segment%HI%jsd,segment%HI%jed
+        if (segment%direction == OBC_DIRECTION_W) then
+          segment%tr_Reg%Tr(nt)%t(I,j,k) = tr_ptr(i+1,j,k)
+        else
+          segment%tr_Reg%Tr(nt)%t(I,j,k) = tr_ptr(i,j,k)
+        endif
+      enddo ; enddo
+    else
+      J=segment%HI%JsdB
+      do k=1,nz ; do i=segment%HI%isd,segment%HI%ied
+        if (segment%direction == OBC_DIRECTION_S) then
+          segment%tr_Reg%Tr(nt)%t(i,J,k) = tr_ptr(i,j+1,k)
+        else
+          segment%tr_Reg%Tr(nt)%t(i,J,k) = tr_ptr(i,j,k)
+        endif
+      enddo ; enddo
+    endif
+    segment%tr_Reg%Tr(nt)%tres(:,:,:) = segment%tr_Reg%Tr(nt)%t(:,:,:)
+  enddo
+!  call setup_OBC_tracer_reservoirs(GV, OBC, .true.) ! Skip T and S to avoid restart bug
```

Comment:
This commented line is a leftover from old debugging and can be deleted. But, there is a necessary addition to this subroutine. The generic tracer segment data was not set up yet during MOM_initialize_state() -> fill_temp_salt_segments() -> setup_obc_tracer_reservoirs(), leaving OBC%tres_* 0 and causing low values to propagate in from the boundary initially. One way to fix this is to copy to OBC%tres_* again in this subroutine. I tried this [here](https://github.com/andrew-c-ross/MOM6/commit/c1acb365eb61f52e23f5bdd6c22758b0a909f192) and it works.